File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
    bool isNotAlphanumeric(char c) {
        return !isalnum(c);
    }
public:
    bool isPalindrome(string s) {
        int i=0, j=s.size() -1;
        while(i<j) {
            if(isNotAlphanumeric(s[i])) i++;
            else if(isNotAlphanumeric(s[j])) j--;
            else if(tolower(s[i]) == tolower(s[j])) i++,j--;
            else return false;
        }
        return true;
    }
};

File: Data Structures & Algorithms/palindrome-partitioning/submission-0.java
 class Solution {
    List<List<String>> res = new ArrayList<>();
    List<String> list = new ArrayList<>();
    private boolean isPalindrome(String s, int i, int j) {
        while(i <= j) {
            if(s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
    private void partition(String s, int begin) {
        if(begin >= s.length()) {
            res.add(new ArrayList<>(list));
        }
        for(int i = begin; i < s.length(); i++) {
            if(isPalindrome(s, begin, i)) {
                list.add(s.substring(begin, i+1));
                partition(s, i+1);
                list.remove(list.size() - 1);
            }
        }
    }

    public List<List<String>> partition(String s) {
            partition(s, 0);
            return res;
        }
}
